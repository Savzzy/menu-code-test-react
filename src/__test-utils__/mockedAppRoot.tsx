import { render, RenderResult } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { ThemeProvider } from "styled-components";
import globalTheme from "../globalTheme";
import rootReducer from "../reducers";
import { intialStore } from "../store";
import { Store } from "../types";

interface MockedAppRootOptions {
  mockStore: Store;
}

const mockedAppRoot = (
  ComponentTree: JSX.Element,
  options?: MockedAppRootOptions,
): RenderResult => {
  const AppContainer: React.FC = ({ children }) => (
    <Provider
      store={createStore(rootReducer, options?.mockStore || intialStore)}
    >
      <ThemeProvider theme={globalTheme}>{children}</ThemeProvider>
    </Provider>
  );

  return render(ComponentTree, { wrapper: AppContainer });
};

export default mockedAppRoot;
