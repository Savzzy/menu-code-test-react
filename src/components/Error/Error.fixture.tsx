import { Store, Diners } from "../../types";
import { UI_ERRORS } from "../../constants";

export const mockStore: Store = {
  activeDiner: Diners.diner1,
  order: {},
  error: UI_ERRORS.DUPLICATE_ORDER_ITEM,
};
