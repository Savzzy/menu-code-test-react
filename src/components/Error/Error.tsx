import React from "react";
import { useSelector, useDispatch } from "react-redux";
import styled from "styled-components";
import { Store } from "../../types";
import { updateErrorState } from "../../actions";

const ErrorContainer = styled.div`
  width: fit-content;
  border-radius: 5px;
  background-color: ${(props) => props.theme.colors.primary};
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 20px;

  position: fixed;
  bottom: 1%;
  right: 1%;
`;

const ErrorMessage = styled.div`
  color: ${(props) => props.theme.colors.lightText};
  font-size: 20px;
  font-weight: 400;
`;

const Error: React.FC = () => {
  const dispatch = useDispatch();
  const error = useSelector((store: Store) => store.error);

  if (!error) {
    return null;
  }

  setTimeout(() => {
    dispatch(updateErrorState(null));
  }, 3000);

  return (
    <ErrorContainer>
      <ErrorMessage>{error}</ErrorMessage>
    </ErrorContainer>
  );
};

export default Error;
