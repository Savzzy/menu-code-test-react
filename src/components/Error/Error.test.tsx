import React from "react";
import Error from "./Error";
import mockedAppRoot from "../../__test-utils__/mockedAppRoot";
import { mockStore } from "./Error.fixture";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: () => {
    return mockStore.error;
  },
}));

describe("Error component", () => {
  it("renders correctly", () => {
    const { container } = mockedAppRoot(<Error />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows the error present in store", () => {
    const { getByText } = mockedAppRoot(<Error />);

    expect(getByText(mockStore.error)).toBeInTheDocument();
  });
});
