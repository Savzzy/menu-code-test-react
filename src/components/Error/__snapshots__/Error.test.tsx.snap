// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error component renders correctly 1`] = `
<div
  class="sc-AxjAm eQJdLU"
>
  <div
    class="sc-AxirZ gZHoxL"
  >
    You cannot order the same item more than once.
  </div>
</div>
`;
